sdl2promela/translator.py

```diff
@@ -296,36 +386,498 @@ def __generate_statement(
     transition: sdlmodel.Transition,
     assignment: sdlmodel.AssignmentTask,
 ) -> promelamodel.Statement:
-    return (
-        AssignmentBuilder()
-        .withTarget(
-            VariableReferenceBuilder(
-                __get_variable_name(sdl_model, assignment.assignment.left)
-            ).build()
-        )
-        .withSource(
-            __generate_statement(sdl_model, transition, assignment.assignment.right)
+    statements: List[promelamodel.Statement] = __generate_assignment(
+        sdl_model,
+        assignment.assignment.left,
+        assignment.assignment.right,
+        assignment.type,
+    )
+    if len(statements) == 1:
+        return statements[0]
+    else:
+        return (
+            BlockBuilder(promelamodel.BlockType.BLOCK)
+            .withStatements(statements)
+            .build()
         )
+
+
+@dispatch(
+    sdlmodel.Model,
+    (
+        sdlmodel.VariableReference,
+        sdlmodel.ArrayAccess,
+        sdlmodel.MemberAccess,
+    ),
+    sdlmodel.Expression,
+    type,
+)
+def __generate_assignment(
+    sdl_model: sdlmodel.Model,
+    left: Union[
+        sdlmodel.VariableReference, sdlmodel.ArrayAccess, sdlmodel.MemberAccess
+    ],
+    right: sdlmodel.Expression,
+    left_type: type,
+) -> List[promelamodel.Statement]:
+    finalType = resolve_asn1_type(sdl_model.types, left_type)
+
+    assignInlineName = __get_assign_value_inline_name(finalType)
+
+    statements: List[promelamodel.Statement] = []
+    statements.append(
+        CallBuilder()
+        .withTarget(assignInlineName)
+        .withParameter(__generate_variable_name(sdl_model, left, True))
+        .withParameter(__generate_expression(sdl_model, right))
         .build()
     )
 
+    return statements
 
-@dispatch(sdlmodel.Model, sdlmodel.Transition, sdlmodel.VariableReference)
-def __generate_statement(
+
+@dispatch(
+    sdlmodel.Model,
+    (
+        sdlmodel.VariableReference,
+        sdlmodel.ArrayAccess,
+        sdlmodel.MemberAccess,
+    ),
+    sdlmodel.EmptyStringValue,
+    type,
+)
+def __generate_assignment(
     sdl_model: sdlmodel.Model,
-    transition: sdlmodel.Transition,
-    variable: sdlmodel.VariableReference,
-) -> promelamodel.Statement:
-    return VariableReferenceBuilder(__get_variable_name(sdl_model, variable)).build()
+    left: Union[
+        sdlmodel.VariableReference, sdlmodel.ArrayAccess, sdlmodel.MemberAccess
+    ],
+    right: sdlmodel.EmptyStringValue,
+    left_type: type,
+) -> List[promelamodel.Statement]:
+    finalType = resolve_asn1_type(sdl_model.types, left_type)
+    if finalType.kind == "OctetStringType" or finalType.kind == "IA5StringType":
```

Comment:
I wonder whether we could provide constants for these strings - data, length, OctetStringType and IA5StringType - what do you think?

Reply:
Done.